### PR TITLE
feat(json): add support of JSON.RESP command

### PR DESCRIPTION
**Pull request by jackyyyyyssss**
https://redis.io/docs/latest/commands/json.resp/
Local testing of multi-layer nested JSON results 6666 kvrocks and 6388 redis stack
image

image
image
image

### CODE REVIEW

File: src/types/json.h [hunk 1]
```diff
@@ -43,7 +43,14 @@
...
 using Optionals = std::vector<std::optional<T>>;
-
+struct ArrayOrObjectPlaceHolder {};
+using VariantType = std::variant<std::string, int64_t, ArrayOrObjectPlaceHolder, bool>;
+template <typename T>
+struct JsonResp {
+  std::vector<std::unique_ptr<JsonResp<VariantType>>> children;
+  T value;
+  explicit JsonResp(T value) : value(std::move(value)) {}
+};
```
Comment (PragmaTwice) [inline src/types/json.h]: We can remove all of these types and just dump the RESP string directly.

**Review by PragmaTwice — CHANGES_REQUESTED** [review]
See above.

File: src/types/json.h [hunk 2]
```diff
@@ -38,12 +38,17 @@
...
 #include "storage/redis_metadata.h"
 
 template <class T>
 using Optionals = std::vector<std::optional<T>>;
-
+struct JsonResp {
+  Optionals<JsonResp> children;
+  std::string value;
+  explicit JsonResp(std::string value) : value(std::move(value)) {}
+};
```
Comment (PragmaTwice) [inline src/types/json.h]: No need. just string is enough here.

Comment (jackyyyyyssss) [reply src/types/json.h]: done PTAL thks

File: src/types/json.h [hunk 3]
```diff
@@ -628,6 +630,57 @@ struct JsonValue {
...
+      json_resps = json_resps + conn->NilString();
+
+    } else if (origin.is_array()) {
+      json_resps = json_resps + redis::MultiLen(origin.size() + 1);
+      json_resps = json_resps + redis::SimpleString("[");
+
+      for (const auto &json_array_value : origin.array_range()) {
+        TransformResp(json_array_value, json_resps, conn);
+      }
+    }
```
Comment (git-hulk) [inline src/types/json.h]:
```suggestion
    if (origin.is_object()) {
      json_resps = json_resps + redis::MultiLen(origin.size() * 2 + 1);
      json_resps = json_resps + redis::SimpleString("{");

      for (const auto &json_kv : origin.object_range()) {
        json_resps = json_resps + redis::BulkString(json_kv.key());
        TransformResp(json_kv.value(), json_resps, conn);
      }

    } else if (origin.is_int64() || origin.is_uint64()) {
      json_resps = json_resps + redis::Integer(origin.as_integer<int64_t>());
    } else if (origin.is_double() || origin.is_string()) {
      json_resps = json_resps + redis::BulkString(origin.as_string());
    } else if (origin.is_bool()) {
      json_resps = json_resps + redis::SimpleString(std::string(origin.as_bool() ? "true" : "false"));
    } else if (origin.is_null()) {
      json_resps = json_resps + conn->NilString();
    } else if (origin.is_array()) {
      json_resps = json_resps + redis::MultiLen(origin.size() + 1);
      json_resps = json_resps + redis::SimpleString("[");
      for (const auto &json_array_value : origin.array_range()) {
        TransformResp(json_array_value, json_resps, conn);
      }
    }
```

File: src/types/json.h [hunk 4]
```diff
@@ -628,6 +630,57 @@ struct JsonValue {
...
+      }
+
+    } else if (origin.is_int64() || origin.is_uint64()) {
+      json_resps = json_resps + redis::Integer(origin.as_integer<int64_t>());
+
+    } else if (origin.is_double()) {
+      json_resps = json_resps + redis::BulkString(origin.as_string());
+
+    } else if (origin.is_string()) {
+      json_resps = json_resps + redis::BulkString(origin.as_string());
```
Comment (PragmaTwice) [inline src/types/json.h]: please use `+=` instead of `= .. + ..`

File: src/types/json.h [hunk 5]
```diff
@@ -628,6 +630,57 @@ struct JsonValue {
     }
     return status;
   }
+  static void TransformResp(const jsoncons::json &origin, std::string &json_resps, redis::Connection *conn) {
```
Comment (PragmaTwice) [inline src/types/json.h]: You can pass RESP instead of Connection after #2406 is merged.

Comment (jackyyyyyssss) [reply src/types/json.h]: Thank you for your guidance in your spare time. I will do so this afternoon @git-hulk @PragmaTwice

File: src/types/json.h [hunk 6]
```diff
@@ -628,6 +629,53 @@ struct JsonValue {
...
+      json_resps += redis::MultiLen(origin.size() + 1);
+      json_resps += redis::SimpleString("[");
+
+      for (const auto &json_array_value : origin.array_range()) {
+        TransformResp(json_array_value, json_resps, resp);
+      }
+    }
+  }
+
+  StatusOr<std::string> JsonConvertResp(std::string_view path, redis::RESP resp) const {
```
Comment (PragmaTwice) [inline src/types/json.h]:
```suggestion
  StatusOr<std::string> ConvertToResp(std::string_view path, redis::RESP resp) const {
```

File: src/types/json.h [hunk 7]
```diff
@@ -628,6 +629,53 @@ struct JsonValue {
...
+        TransformResp(json_kv.value(), json_resps, resp);
+      }
+
+    } else if (origin.is_int64() || origin.is_uint64()) {
+      json_resps += redis::Integer(origin.as_integer<int64_t>());
+
+    } else if (origin.is_string() || origin.is_double()) {
+      json_resps += redis::BulkString(origin.as_string());
+    } else if (origin.is_bool()) {
+      json_resps += redis::SimpleString(std::string(origin.as_bool() ? "true" : "false"));
```
Comment (PragmaTwice) [inline src/types/json.h]:
```suggestion
      json_resps += redis::SimpleString(origin.as_bool() ? "true" : "false");
```

File: src/types/json.h [hunk 8]
```diff
@@ -628,6 +629,53 @@ struct JsonValue {
...
+        query_count++;
+        TransformResp(origin, json_resps, resp);
+      });
+    } catch (const jsoncons::jsonpath::jsonpath_error &e) {
+      return {Status::NotOK, e.what()};
+    }
+    if (path != "$") {
+      return redis::MultiLen(query_count) += json_resps;
+    }
+    return json_resps;
```
Comment (PragmaTwice) [inline src/types/json.h:677]: seems `query_count` is useless

File: src/types/json.h [hunk 9]
```diff
@@ -628,6 +629,53 @@ struct JsonValue {
...
+      jsoncons::jsonpath::json_query(value, path, [&](const std::string & /*path*/, const jsoncons::json &origin) {
+        query_count++;
+        TransformResp(origin, json_resps, resp);
+      });
+    } catch (const jsoncons::jsonpath::jsonpath_error &e) {
+      return {Status::NotOK, e.what()};
+    }
+    if (path != "$") {
+      return redis::MultiLen(query_count) += json_resps;
+    }
```
Comment (PragmaTwice) [inline src/types/json.h]:
```suggestion
      return redis::MultiLen(query_count) + json_resps;
    }
```
?

Comment (jackyyyyyssss) [reply src/types/json.h]: fix thanks

File: src/types/json.h [hunk 10]
```diff
@@ -628,6 +629,53 @@ struct JsonValue {
...
+      jsoncons::jsonpath::json_query(value, path, [&](const std::string & /*path*/, const jsoncons::json &origin) {
+        query_count++;
+        TransformResp(origin, json_resps, resp);
+      });
+    } catch (const jsoncons::jsonpath::jsonpath_error &e) {
+      return {Status::NotOK, e.what()};
+    }
+    if (path != "$") {
+      return redis::MultiLen(query_count) + json_resps;
+    }
```
Comment (PragmaTwice) [inline src/types/json.h]: seems not consistent with RedisJSON here.

`json.resp a $` and `json.resp a` have different behavior.

Comment (PragmaTwice) [reply src/types/json.h]: we can move `redis::MultiLen(query_count)` here to `CommandJsonResp::Parse/Execute`.

and here we can return `vector<string>`.

Comment (jackyyyyyssss) [reply src/types/json.h]: Thank you for the reminder. It has not been modified yet. make the PR draft first

Comment (jackyyyyyssss) [reply src/types/json.h]: It has been modified, but it is a bit unsupported. jsoncons does not support queries without $, Redis 7.0 supports it and will remove the outermost array

**Review by PragmaTwice — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Pull request merged**